github-runner-manager/src/github_runner_manager/manager/runner_manager.py

```diff
@@ -541,51 +468,180 @@ def _create_runner(args: _CreateRunnerArgs) -> InstanceID:
-        return False
-
-    # Always delete deletable runners with health information.
-    if health.deletable:
-        return True
+    # Always clean all runners in the platform that are not in the cloud
+    dangling_runners: set[str] = set(
+        runner.metadata.runner_id
+        for runner in runners.non_requested_runners
+        if runner.metadata.runner_id
+    )
+    logger.info("Dangling runners IDs: %s", dangling_runners)
+
+    deletable_runners: set[str] = set(
+        runner.identity.metadata.runner_id
+        for runner in runners.requested_runners
+        if runner.deletable and runner.identity.metadata.runner_id
+    )
+    logger.info("Deletable runner IDs: %s", deletable_runners)
+
+    vm_instance_id_map = {vm.instance_id: vm for vm in vms}
+    # Kill old runners that are offline and idle as they could be in failed state.
+    # We may also kill here runners that were online and not busy and went temporarily to
+    # offline, but that should not be an issue, as those runners will be spawned again.
+    timed_out_offline_idle_runners: set[str] = set(
+        runner.identity.metadata.runner_id
+        for runner in runners.requested_runners
+        if runner.identity.metadata.runner_id
+        and not runner.online
+        and not runner.busy
+        and runner.identity.instance_id in vm_instance_id_map
+        and vm_instance_id_map[runner.identity.instance_id].is_older_than(
+            RUNNER_MAXIMUM_CREATION_TIME
+        )
+    )
+    logger.info("Timed out offline idle runner IDs: %s", timed_out_offline_idle_runners)
+
+    online_idle_runners: set[str] = set(
+        runner.identity.metadata.runner_id
+        for runner in runners.requested_runners
+        if clean_idle and runner.online and not runner.busy and runner.identity.metadata.runner_id
+    )
+    logger.info("Online idle runners (clean idle: %s): %s", clean_idle, online_idle_runners)
+
+    runners_to_cleanup = list(
+        dangling_runners | deletable_runners | timed_out_offline_idle_runners | online_idle_runners
+    )
+    if not force_delete:
+        return runners_to_cleanup
+
+    if max_runners_to_delete is None:
+        logger.info("Force delete all runners")
+        return [
+            runner.metadata.runner_id
+            for runner in runners.non_requested_runners
+            + [runner.identity for runner in runners.requested_runners]
+            if runner.metadata.runner_id
+        ]
 
-    if clean_idle and health.online and not health.busy:
-        return True
+    num_runners_to_cleanup = len(runners_to_cleanup)
+    # max_runners_to_delete is used for scaling down - if we're already able to clean up enough
+    # runners, skip scaling down.
+    # this function is targeted for refactor in a follow up PR to separate scaling logic from
+    # cleanup logic.
+    if num_runners_to_cleanup >= max_runners_to_delete:
+        return runners_to_cleanup
+
+    max_runners_to_delete -= num_runners_to_cleanup
```

Comment:
nit: I usually prefer to not modify variables, especially when the function scope is really big,  I'd prefer adding a new variable `max_remaining_runners_to_delete`